### DISCUSSION TITLE
Logging results/output of backups and checks on Windows 10

### DISCUSSION THREAD

**Discussion by ScottyDoesKnow** [erikw/restic-automatic-backup-scheduler#126] [Q&A] [answered]
I have an application running on my PC that checks various things including my current backup solution so there's no need to send emails or anything. But like my current backup solution, I'd like to have a log to check to make sure it ran successfully and to check for errors. I first tried using RESTIC_NOTIFY_BACKUP_STATS and RESTIC_BACKUP_NOTIFICATION_FILE, but you don't get much information on success and it doesn't seem to produce anything on failure. Even if I just had one file that was write-locked it would complete the snapshot but not produce the file.

I figured out how to get the output of restic_backup.sh by modifying the scheduled task action:

```
Program: "C:\Program Files\Git\git-bash.exe"
Arguments: -l -c "source /etc/restic/default.env.sh && /bin/restic_backup.sh >> /c/backup/restic_backup.log 2>&1"
```

But then I tried doing the same with restic_check.sh and this was my output:

`restic check --option b2.connections=10 --verbose=1`

Looking at the script, it seems to be because it uses echo to run the command.

Also, it would be great if it didn't open a bash window when the scheduled task ran. I'm considering just running everything through my application with C# Process to try to make this work but I know it's gonna be a fun time passing and escaping all the arguments correctly. So if anyone knows a better way, that would be great.

**Comment by erikw** [comment]
Not having the terminal window pop up during backup would be useful, if the solution is simple and doesn’t add dependencies. I’d like to not add too much functionality at this point given the https://github.com/erikw/restic-automatic-backup-scheduler#project-scope :)

**Reply by ScottyDoesKnow** [reply]
Thanks so much for all the prompt replies, and all the work you've done on this! It "should" be relatively easy to make a powershell script that does it without a window, but I also remember wasting hours trying to automate a Visual Studio install due to issues passing/escaping parameters. If I figured it out and made a powershell script that hid the window and potentially took an argument for a log directory, would you want to include it?

Also, it seems like the echo in restic_check.sh line 59 might just be a mistake? Because it prints the command instead of running it, removing the echo made it work. Took me a while to figure out because I was editing the wrong restic_check.sh, which is probably why my attempt to change directories to an empty one in the backup script didn't do anything. But I'll wait until my backup is done to test that.

**Comment by erikw** [comment, marked as answer]
> If I figured it out and made a powershell script that hid the window and potentially took an argument for a log directory, would you want to include it?

It's hard to say until seeing what what means ¯\_(ツ)_/¯ . The intention of this repo is to be a base for further hacking to make a personal perfect backup solution. So if you have the motivation, aim to implement this for yourself as your main goal. Then if suitable (small, low complexity, not adding dependencies) it could be added here too :) 


> Also, it seems like the echo in restic_check.sh line 59 might just be a mistake? 

Yes indeed! Removed in 48ec2f9cf7f3ed4d8733fd5683fb3a87826a0e1f

**Comment by ScottyDoesKnow** [comment]
So I've got a script but it's not perfect. It flickers up a powershell window for a split second, but at least nothing stays open. You can prevent that by setting it to run whether the user is logged on or not, which seems to work for me but I'm not sure I'd trust it as a default. The main issue though is that I'm not sure the install script could handle it cleanly. Here's the example usage:

Program: powershell.exe
Arguments: -WindowStyle Hidden -File "C:\backup\restic_task.ps1" -script restic_backup.sh
or
Arguments: -WindowStyle Hidden -File "C:\backup\restic_task.ps1" -script restic_check.sh

And here's what the script is doing:

```
#!/usr/bin/env pwsh
# Runs the given restic-automatic-backup-scheduler bash script with the given arguments
# -script: which script to run (probably restic_backup.sh or restic_check.sh)
# -envs: array of environment scripts to run the given bash script with
# -bashPath: path to bash.exe
# -logDirectory: directory to output logs to

param(
    [string]$script,
    [string[]]$envs = @("default.env.sh"),
    [string]$bashPath = "C:\Program Files\Git\bin\bash.exe",
    [string]$logDirectory
)

if ([string]::IsNullOrWhiteSpace($script)) {
    Write-Error "-script is a required parameter."
    exit 1
}

# Because passing parameters is fun!
if ($envs.Count -eq 1 -and $envs[0].Contains(',')) {
    $envs = $envs[0].Split(',')
}

foreach ($env in $envs) {
    if (![string]::IsNullOrWhiteSpace($logDirectory)) {
        if (!(Test-Path -Path $logDirectory -PathType Container)) {
            Write-Error "The specified -logDirectory does not exist: $logDirectory"
            exit 1
        }
    
        $timestamp = Get-Date -Format "yyyyMMdd-HHmmss"
        $scriptWithoutExtension = [System.IO.Path]::GetFileNameWithoutExtension($script)
        if ($env -ne "default.env.sh") {
            $envString = "_" + ($env -replace "\..+$")
        } else {
            $envString = ""
        }
        $logFileNameBase = Join-Path $logDirectory ("$scriptWithoutExtension" + $envString + "_" + "$timestamp")
        $logFilePathRunning = $logFileNameBase + "_running.log"
        $logFilePathFinal = $logFileNameBase + ".log"

        $logFilePathBash = $logFilePathRunning -replace '^([A-Z]):', '/$1' -replace '\\', '/'
        $bashArgs = '-c "source /etc/restic/{0} && /bin/{1} > {2} 2>&1"' -f $env,$script, $logFilePathBash
    } else {
        $bashArgs = '-c "source /etc/restic/{0} && /bin/{1}"' -f $env,$script
    }

    $processStartInfo = New-Object System.Diagnostics.ProcessStartInfo
    $processStartInfo.FileName = $bashPath
    $processStartInfo.Arguments = $bashArgs
    $processStartInfo.UseShellExecute = $false
    $processStartInfo.CreateNoWindow = $true

    $process = New-Object System.Diagnostics.Process
    $process.StartInfo = $processStartInfo
    $processStarted = $process.Start()

    if ($processStarted) {
        Write-Output "bash process running $script with $env..."
    } else {
        Write-Error "bash process failed to start."
        exit 1
    }

    $process.WaitForExit()

    $bashFailed = $false
    if ($process.ExitCode -eq 0) {
        Write-Output "bash process completed successfully."
    } else {
        Write-Output ("bash process failed with exit code: " + $process.ExitCode)
        $bashFailed = $true
    }

    if (![string]::IsNullOrWhiteSpace($logFilePathRunning)) {
        if (!(Test-Path -Path $logFilePathRunning -PathType Leaf)) {
            Write-Error "The specified file does not exist: $logFilePathRunning"
            exit 1
        }

        Move-Item $logFilePathRunning $logFilePathFinal
    }

    if ($bashFailed) {
        exit 1
    }
}
```

I tried getting the task to fail with the exit codes, but it doesn't seem to want to. It lists it as the last run result but still succeeds.

**Reply by erikw** [reply]
I'll leave this discussion open here in case someone else would drop by with some ideas or knowledge on the Windows side here :)

**Comment by ScottyDoesKnow** [comment]
Ya, there might be some issues with putting this in the project (getting this to work with the install script that creates the scheduled tasks mainly), but I wanted to post it here anyway for other people to find if they need it.
